src/Shared/CommunicationsUtilities.cs

```diff
@@ -330,14 +330,16 @@ internal static long GetHostHandshake(HandshakeOptions nodeType)
 
             Trace("MSBUILDNODEHANDSHAKESALT=\"{0}\", msbuildDirectory=\"{1}\", nodeType={2}, FileVersionHash={3}", salt, toolsDirectory, nodeType, FileVersionHash);
 
-            //FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
-            //nodeType (4 bits) is shifted just after the FileVersionHash
-            //nodeHandshakeSalt (32 bits) is shifted just after hostContext
-            //the most significant byte (leftmost 8 bits) will get zero'd out to avoid connecting to older builds.
-            //| masked out | nodeHandshakeSalt | hostContext |              fileVersionHash             | SessionID
-            //  0000 0000     0000 0000 0000        0000        0000 0000 0000 0000 0000 0000 0000 0000   0000 0000
-            long baseHandshake = ((long)nodeHandshakeSalt << 44) | ((long)nodeType << 40) | ((long)FileVersionHash << 8);
-            return GenerateHostHandshakeFromBase(baseHandshake);
+            // FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
+            // HandshakeOptions (5 bits) is shifted just after the FileVersionHash
+            // remaining bits of nodeHandshakeSalt (32 bits truncated to 19) are shifted next
```

Comment:
nit: It's effectively truncated to 11 bits because the most significant byte is masked out in `GenerateHostHandshakeFromBase()`. And I don't think the session ID is necessarily an 8-bit value either.

Considering the above and taking a step back, I'm wondering why we insist on encoding everything in a single `long` instead of defining a proper back- and forward-compatible structure.

```C#
struct Handshake
{
  byte zeroForBackCompat = 0;
  byte lengthForForwardCompat = sizeof(Handshake);
  NodeType nodeType;
  int handshakeSalt;
  int versionHash;
  int sessionId;
  // ...
}
```

Reply:
That does make logging on the rejecting side easier too. Should we push this to unblock VS insertions and other PRs and then switch to that?

Reply:
I think it would make sense to merge this PR now and track making further improvements as a lower-pri work item.

Reply:
I'd even think about dropping the version _hash_ and just sending the 4 parts of AssemblyFileVersion?

Reply:
@rainersigwald, @ladipro

I put my first draft [here](https://github.com/microsoft/msbuild/compare/master...Forgind:struct-handshake?expand=1).

There are two parts that I think still need fixing:
1) fileVersion is currently a string because the four parts were returned together. @rainersigwald, is there a way to get the parts individually and turn them into ints if necessary, is there a premade way to do that, or do I have to roll my own (probably regex) parser?
2) It's claiming I need to import ValueTuple before I can have `internal IEnumerable<int part, int index)> RetrieveHandshakeComponents()`. Then it's making me say `handshake.RetrieveHandshakeComponents().First().Item1` instead of `handshake.RetrieveHandshakeComponents().First().part` https://github.com/microsoft/msbuild/pull/5270/ made it look like I could just have the valueTuples, and it would work. What do I need to do differently?

Reply:
> fileVersion is currently a string because the four parts were returned together

You should be able to call `Assembly.GetName().Version` to get to the four version parts but it looks like in the current code we use `AssemblyInformationalVersion` which could be an arbitrary string and not necessarily just integers e.g. "16.7.0-dev-20324-01+ee1c9fd0c7e0e43174785a4e2ed177f14f5856a0". Is it maybe too restrictive to have the handshake consist only of integers?

For the new handshake code please consider all these scenarios, both from the client and host perspective:
- The code you are writing trying to connect to a very old MSBuild with an 8-bit handshake.
- The code you are writing trying to connect to a not-so-old MSBuild with a 64-bit handshake.
- The code you are writing trying to connect to a newer MSBuild with an updated `Handshake` structure.

and make sure that the handshake will reliably fail without hanging.